File: BOJ/15683_1_1.cpp
// 감시
// https://www.acmicpc.net/problem/15683
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
const int MAX = 8 + 1;
const int dr[4] = {0, 1, 0, -1};
const int dc[4] = {1, 0, -1, 0};
int N, M, answer = 987654321;
int new_board[MAX][MAX];
int board[MAX][MAX];
int visited[MAX];
vector<pair<int, int>> cctv;

void right(int dir, int s_r, int s_c)
{
	int r = s_r;
	int c = s_c;
	int new_dir = (0 + dir) % 4;

	while (true)
	{
		int nr = r + dr[new_dir];
		int nc = c + dc[new_dir];

		if (nr < 0 || nr >= N || nc < 0 || nc >= M || new_board[nr][nc] == 6)
			break;
		if (new_board[nr][nc] == 0)
			new_board[nr][nc] = -1;
		r = nr;
		c = nc;
	}
}
void down(int dir, int s_r, int s_c)
{
	int r = s_r;
	int c = s_c;
	int new_dir = (1 + dir) % 4;

	while (true)
	{
		int nr = r + dr[new_dir];
		int nc = c + dc[new_dir];

		if (nr < 0 || nr >= N || nc < 0 || nc >= M || new_board[nr][nc] == 6)
			break;
		if (new_board[nr][nc] == 0)
			new_board[nr][nc] = -1;
		r = nr;
		c = nc;
	}
}

void left(int dir, int s_r, int s_c)
{
	int r = s_r;
	int c = s_c;
	int new_dir = (2 + dir) % 4;

	while (true)
	{
		int nr = r + dr[new_dir];
		int nc = c + dc[new_dir];

		if (nr < 0 || nr >= N || nc < 0 || nc >= M || new_board[nr][nc] == 6)
			break;
		if (new_board[nr][nc] == 0)
			new_board[nr][nc] = -1;
		r = nr;
		c = nc;
	}
}
void up(int dir, int s_r, int s_c)
{
	int r = s_r;
	int c = s_c;
	int new_dir = (3 + dir) % 4;

	while (true)
	{
		int nr = r + dr[new_dir];
		int nc = c + dc[new_dir];

		if (nr < 0 || nr >= N || nc < 0 || nc >= M || new_board[nr][nc] == 6)
			break;
		if (new_board[nr][nc] == 0)
			new_board[nr][nc] = -1;
		r = nr;
		c = nc;
	}
}

void check_sight()
{
	for (int r = 0; r < N; r++)
		for (int c = 0; c < M; c++)
			new_board[r][c] = board[r][c];

	for (int i = 0; i < cctv.size(); i++)
	{
		// cctv 번호, index
		int cur_r = cctv[i].first;
		int cur_c = cctv[i].second;
		int cctv_num = board[cur_r][cur_c];
		int dir = visited[i];

		if (cctv_num == 1)
		{
			right(dir, cur_r, cur_c);
		}
		else if (cctv_num == 2)
		{
			//if (dir % 2 != 0)
			//	dir++;
			left(dir, cur_r, cur_c);
			right(dir, cur_r, cur_c);
		}
		else if (cctv_num == 3)
		{
			up(dir, cur_r, cur_c);
			right(dir, cur_r, cur_c);
		}
		else if (cctv_num == 4)
		{
			left(dir, cur_r, cur_c);
			up(dir, cur_r, cur_c);
			right(dir, cur_r, cur_c);
		}
		else
		{
			left(dir, cur_r, cur_c);
			up(dir, cur_r, cur_c);
			right(dir, cur_r, cur_c);
			down(dir, cur_r, cur_c);
		}
	}

	int temp = 0;
	for (int r = 0; r < N; r++)
	{
		for (int c = 0; c < M; c++)
		{
			if (new_board[r][c] == 0)
				temp++;
		}
	}

	if (temp < answer)
		answer = temp;
}

void dfs(int cur)
{
	if (cctv.size() == cur)
	{
		check_sight();
		return;
	}

	for (int dir = 0; dir < 4; dir++)
	{
		visited[cur] = dir;
		dfs(cur + 1);
	}
}

void solve()
{
	dfs(0);
}

void init()
{
	cin >> N >> M;
	for (int r = 0; r < N; r++)
	{
		for (int c = 0; c < M; c++)
		{
			cin >> board[r][c];
			if (board[r][c] >= 1 && board[r][c] <= 5)
			{
				cctv.push_back(make_pair(r, c));
			}
		}
	}
}

void print()
{
	cout << answer << '\n';
}

int main()
{
	ios::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);

	init();
	solve();
	print();
}

File: BOJ/15683_1_1.py
# 감시
# https://www.acmicpc.net/problem/15683
import sys

rotSize = [4, 2, 4, 4, 1]
N, M = map(int, sys.stdin.readline().split())
answer = 987654321
board = []
cctv = []
for r in range(N):
    new_line = list(map(int, sys.stdin.readline().split()))
    for c in range(M):
        if new_line[c] >= 1 and new_line[c] <= 5:
            cctv.append((r, c, new_line[c] - 1))
    board.append(new_line)


def update(dir, cctv):
    global board
    dir = (dir % 4)
    if dir == 0:
        for c in range(cctv[1] + 1, M):
            if board[cctv[0]][c] == 6:
                break
            board[cctv[0]][c] = -1
    elif dir == 1:
        for r in range(cctv[0]-1, -1, -1):
            if board[r][cctv[1]] == 6:
                break
            board[r][cctv[1]] = -1
    elif dir == 2:
        for c in range(cctv[1] - 1, -1, -1):
            if board[cctv[0]][c] == 6:
                break
            board[cctv[0]][c] = -1
    elif dir == 3:
        for r in range(cctv[0] + 1, N):
            if board[r][cctv[1]] == 6:
                break
            board[r][cctv[1]] = -1


def boardCopy(desc, src):
    for r in range(N):
        for c in range(M):
            desc[r][c] = src[r][c]


def dfs(curr):
    global board
    if curr == len(cctv):
        global answer
        sight_cnt = 0
        for r in range(N):
            for c in range(M):
                if board[r][c] == 0:
                    sight_cnt += 1
        answer = min(answer, sight_cnt)
        return

    backup = [[0 for _ in range(M)] for _ in range(N)]
    cctv_type = cctv[curr][2]
    for dir in range(rotSize[cctv_type]):
        boardCopy(backup, board)
        if cctv_type == 0:
            update(dir, cctv[curr])
        elif cctv_type == 1:
            update(dir, cctv[curr])
            update(dir + 2, cctv[curr])
        elif cctv_type == 2:
            update(dir, cctv[curr])
            update(dir + 1, cctv[curr])
        elif cctv_type == 3:
            update(dir, cctv[curr])
            update(dir + 1, cctv[curr])
            update(dir + 2, cctv[curr])
        elif cctv_type == 4:
            update(dir, cctv[curr])
            update(dir + 1, cctv[curr])
            update(dir + 2, cctv[curr])
            update(dir + 3, cctv[curr])

        dfs(curr + 1)
        boardCopy(board, backup)


dfs(0)
print(answer)
